lib/spack/spack/error.py

```diff
@@ -131,10 +131,26 @@ def __init__(self, provided, required=None, constraint_type=None, conflicts=None
             super(UnsatisfiableSpecError, self).__init__(
                 "%s does not satisfy %s" % (provided, required))
         else:
+            import spack.solver.asp as solver  # avoid circular import
+
```

Comment:
Can we avoid this hidden dependency by adding a 5th argument to `__init__`? Idea would be to accept a callable that returns the message we want to print. In this way the function is injected by the caller when raising, which is already in `spack.solver.asp`.

Reply:
Added a new commit, different solution than you suggested